Valid class fails with "java.lang.ClassFormatError" during plugin application

### Current Behavior

I'm maintaining a gradle plugin that contains the following class:
```java
public abstract class RuleClosure extends Closure<Task> {

    protected RuleClosure(Object owner) {
        super(Require.nonNull(owner));
    }

    public abstract Task doCall(String taskName);
}
```

the compiled bytecode with `javap -v` looks like this:
```
{
  protected org.moe.gradle.groovy.closures.RuleClosure(java.lang.Object);
    descriptor: (Ljava/lang/Object;)V
    flags: (0x0004) ACC_PROTECTED
    Code:
      stack=2, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokestatic  #1                  // Method org/moe/gradle/utils/Require.nonNull:(Ljava/lang/Object;)Ljava/lang/Object;
         5: invokespecial #7                  // Method groovy/lang/Closure."<init>":(Ljava/lang/Object;)V
         8: return
      LineNumberTable:
        line 29: 0
        line 30: 8
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       9     0  this   Lorg/moe/gradle/groovy/closures/RuleClosure;
            0       9     1 owner   Ljava/lang/Object;

  public abstract org.gradle.api.Task doCall(java.lang.String);
    descriptor: (Ljava/lang/String;)Lorg/gradle/api/Task;
    flags: (0x0401) ACC_PUBLIC, ACC_ABSTRACT
}
```
`doCall` contains no "Code" section, since it is abstract.
However, when trying to import a plugin with this class I get the following exception:
```java
Caused by: java.lang.ClassFormatError: Code attribute in native or abstract methods in class file org/moe/gradle/groovy/closures/RuleClosure
	at org.gradle.internal.classloader.VisitableURLClassLoader$InstrumentingVisitableURLClassLoader.findClass(VisitableURLClassLoader.java:186)
	at org.gradle.internal.classloader.VisitableURLClassLoader$InstrumentingVisitableURLClassLoader.findClass(VisitableURLClassLoader.java:186)
```

### Expected Behavior

The class is seemingly valid and should therefor not throw an exception.

### Context (optional)

_No response_

### Steps to Reproduce

1. `git clone https://github.com/Berstanio/TestMOE`
2. `./gradlew build` (or related)

### Gradle version

8.4+

### Gradle version that used to work

8.3

### Build scan URL (optional)

_No response_

### Your Environment (optional)

I tested it on multiple JDK's with java 11 and 17.

--- a/subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/ClosureInstrumentationIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/ClosureInstrumentationIntegrationTest.groovy
@@ -1,0 +1,82 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
+
+class ClosureInstrumentationIntegrationTest extends AbstractIntegrationSpec {
+    @Issue("https://github.com/gradle/gradle/issues/28389")
+    def "can instrument abstract closure classes"() {
+        given:
+        createDir("buildSrc") {
+            dir("src/main/java") {
+                file("MyBaseClosure.java") << """
+                    import ${Closure.name};
+                    public abstract class MyBaseClosure extends Closure<String> {
+                        protected MyBaseClosure(Object owner) {
+                            super(owner);
+                        }
+
+                        protected final void setDelegateBase(Object delegate) {
+                            super.setDelegate(delegate);
+                        }
+
+                        @Override
+                        public abstract void setDelegate(Object delegate);
+
+                        public abstract String doCall(String argument);
+                    }
+                """
+
+                file("ReverseClosure.java") << """
+                    public class ReverseClosure extends MyBaseClosure {
+                        protected ReverseClosure(Object owner) {
+                            super(owner);
+                        }
+
+                        @Override
+                        public void setDelegate(Object delegate) {
+                            setDelegateBase(delegate);
+                        }
+
+                        @Override
+                        public String doCall(String argument) {
+                            return new StringBuilder(argument).reverse().toString();
+                        }
+                    }
+                """
+            }
+            file("settings.gradle") << "\n"
+        }
+
+        buildFile("""
+            task("hello") {
+                doLast {
+                    def cl = new ReverseClosure(this)
+                    println("closure = " + cl("hello"))
+                }
+            }
+        """)
+
+        when:
+        run("hello")
+
+        then:
+        outputContains("closure = olleh")
+    }
+}

--- a/subprojects/core/src/main/java/org/gradle/internal/classpath/CallInterceptionClosureInstrumentingClassVisitor.java
+++ b/subprojects/core/src/main/java/org/gradle/internal/classpath/CallInterceptionClosureInstrumentingClassVisitor.java
@@ -107,10 +107,11 @@ public class CallInterceptionClosureInstrumentingClassVisitor extends ClassVisit
          * Renames the Closure's original `doCall` method and adds a wrapper method that invokes the original one.
          */
         RENAME_ORIGINAL_DO_CALL("doCall", null, false, (clazz, methodData) -> {
-            boolean isDoCallMethod = methodData.name.equals("doCall");
-            String methodNameToVisit = isDoCallMethod ? "doCall$original" : methodData.name;
+            // A Closure implementation may have an abstract doCall method. It makes no sense to rewrite that.
+            boolean isValidDoCallMethod = !methodData.isAbstract() && methodData.name.equals("doCall");
+            String methodNameToVisit = isValidDoCallMethod ? "doCall$original" : methodData.name;
             MethodVisitor original = clazz.visitor.visitMethod(methodData.access, methodNameToVisit, methodData.descriptor, methodData.signature, methodData.exceptions);
-            if (isDoCallMethod) {
+            if (isValidDoCallMethod) {
                 @NonNullApi
                 class MethodVisitorScopeImpl extends MethodVisitorScope {
                     public MethodVisitorScopeImpl(MethodVisitor methodVisitor) {
@@ -229,6 +230,10 @@ public class CallInterceptionClosureInstrumentingClassVisitor extends ClassVisit
                 this.descriptor = descriptor;
                 this.signature = signature;
                 this.exceptions = exceptions;
+            }
+
+            public boolean isAbstract() {
+                return (access & Opcodes.ACC_ABSTRACT) != 0;
             }
         }
 

--- a/subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java
+++ b/subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java
@@ -70,7 +70,7 @@ public class InstrumentingClassTransform implements ClassTransform {
     /**
      * Decoration format. Increment this when making changes.
      */
-    private static final int DECORATION_FORMAT = 35;
+    private static final int DECORATION_FORMAT = 36;
 
     private static final Type SYSTEM_TYPE = getType(System.class);
     private static final Type INTEGER_TYPE = getType(Integer.class);

--- a/subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/internal/classpath/DefaultCachedClasspathTransformerTest.groovy
@@ -272,7 +272,7 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jar(file)
         def classpath = DefaultClassPath.of(file)
-        def cachedFile = testDir.file("cached/46be64fb645b6862b6ca71ccee62da52/thing.jar")
+        def cachedFile = testDir.file("cached/81ce855514a44b94a5a32d6f9040cc34/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -301,7 +301,7 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def dir = testDir.file("thing.dir")
         classesDir(dir)
         def classpath = DefaultClassPath.of(dir)
-        def cachedFile = testDir.file("cached/0b371d5ebfeeba14a4cc4903b48dee77/thing.dir")
+        def cachedFile = testDir.file("cached/44eba99864a0061bf21b4b8b91235288/thing.dir")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -332,8 +332,8 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jar(file)
         def classpath = DefaultClassPath.of(dir, file)
-        def cachedDir = testDir.file("cached/0b371d5ebfeeba14a4cc4903b48dee77/thing.dir")
-        def cachedFile = testDir.file("cached/46be64fb645b6862b6ca71ccee62da52/thing.jar")
+        def cachedDir = testDir.file("cached/44eba99864a0061bf21b4b8b91235288/thing.dir")
+        def cachedFile = testDir.file("cached/81ce855514a44b94a5a32d6f9040cc34/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -390,8 +390,8 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file3 = testDir.file("thing3.jar")
         jar(file3)
         def classpath = DefaultClassPath.of(dir, file, dir2, file2, dir3, file3)
-        def cachedDir = testDir.file("cached/0b371d5ebfeeba14a4cc4903b48dee77/thing.dir")
-        def cachedFile = testDir.file("cached/46be64fb645b6862b6ca71ccee62da52/thing.jar")
+        def cachedDir = testDir.file("cached/44eba99864a0061bf21b4b8b91235288/thing.dir")
+        def cachedFile = testDir.file("cached/81ce855514a44b94a5a32d6f9040cc34/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)
@@ -445,7 +445,7 @@ class DefaultCachedClasspathTransformerTest extends ConcurrentSpec {
         def file = testDir.file("thing.jar")
         jarWithStoredResource(file)
         def classpath = DefaultClassPath.of(file)
-        def cachedFile = testDir.file("cached/9887cf59e55e07a7da0f3ad17dbea9fa/thing.jar")
+        def cachedFile = testDir.file("cached/d6067f6fe41da98247e6c867585958b2/thing.jar")
 
         when:
         def cachedClasspath = transformer.transform(classpath, BuildLogic)

